src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -3827,62 +3821,172 @@ VOID MethodTableBuilder::AllocateFieldDescs()
+    {
+        if (pModuleOfTypeDef->IsSystem())
+        {
+            if (tkTypeDefExtends == CoreLibBinder::GetClassIfExist(CLASS__ENUM)->GetCl())
+            {
+                thisIsAnEnum = true;
+            }
+        }
+    }
+    else
+    {
+        _ASSERTE(TypeFromToken(tkTypeDefExtends) == mdtTypeRef);
+        LPCSTR pszNameSpace;
+        LPCSTR pszClassName;
+        pModuleOfTypeDef->GetMDImport()->GetNameOfTypeRef(tkTypeDefExtends, &pszNameSpace, &pszClassName);
+        if (pszNameSpace != NULL && pszClassName != NULL && strcmp(pszNameSpace, "System") == 0 && strcmp(pszClassName, "Enum") == 0)
+        {
+            Module *pModuleOfSystemEnumType;
+            mdTypeDef tkTypeDefOfSystemEnumType;
 
-    return MetaSig::CompareElementType(pFakeSig, pFieldSig,
-                                       pFakeSig + cFakeSig,  pMemberSignature + cMemberSignature,
-                                       GetModule(), GetModule(),
-                                       NULL, NULL);
+            ClassLoader::ResolveTokenToTypeDefThrowing(pModuleOfTypeDef, tkTypeDefExtends, &pModuleOfSystemEnumType, &tkTypeDefOfSystemEnumType);
 
-}
+            if (pModuleOfSystemEnumType != NULL && pModuleOfSystemEnumType->IsSystem())
+            {
+                thisIsAnEnum = true;
+            }
+        }
+    }
 
-//*******************************************************************************
-//
-// Used pByValueClass cache to mark self-references
-//
-static BOOL IsSelfRef(MethodTable * pMT)
-{
-    return pMT == (MethodTable *)-1;
+    if (thisIsAnEnum)
+    {
+        // If this is an enum, we can return the type of the instance field
+        IMDInternalImport *pEnumModuleImport = pModuleOfTypeDef->GetMDImport();
+        HENUMInternalHolder hEnumField(pEnumModuleImport);
+        IfFailThrow(hEnumField.EnumInitNoThrow(mdtFieldDef, tkTypeDef));
+        mdFieldDef tkFieldDef;
+        while (hEnumField.EnumNext(&tkFieldDef))
+        {
+            ULONG cMemberSignature;
+            PCCOR_SIGNATURE pMemberSignature;
+            DWORD flags;
+
+            IfFailThrow(pEnumModuleImport->GetFieldDefProps(tkFieldDef, &flags));
+
+            if (IsFdStatic(flags) || IsFdLiteral(flags))
+            {
+                continue; // Skip static and literal fields
+            }
+            IfFailThrow(pEnumModuleImport->GetSigOfFieldDef(tkFieldDef, &cMemberSignature, &pMemberSignature));
+
+            SigParser sig(pMemberSignature, cMemberSignature);
+            uint32_t ulCallConv;
+            IfFailThrow(sig.GetCallingConvInfo(&ulCallConv));
+            if (ulCallConv != IMAGE_CEE_CS_CALLCONV_FIELD)
+            {
+                ThrowHR(COR_E_TYPELOAD);
+            }
+
+            CorElementType elemType;
+            IfFailThrow(sig.GetElemType(&elemType));
+
+            switch (elemType)
+            {
+                case ELEMENT_TYPE_BOOLEAN:
+                case ELEMENT_TYPE_CHAR:
```

Comment:
> Note: The char type cannot be used as an underlying type, either by keyword or via an integral_type_name. end note

Reply:
The ECMA says:
```
The symbols of an enum are represented by an underlying integer type: one of { bool, char, int8, unsigned int8, int16, unsigned int16, int32, unsigned int32, int64, unsigned int64, native int, unsigned native int } 
```
but CoreCLR also accepts float and double as tested in [1](https://github.com/dotnet/runtime/blob/main/src/tests/JIT/Intrinsics/TypeIntrinsicsEnums.il) [2](https://github.com/dotnet/runtime/blob/main/src/tests/JIT/opt/Devirtualization/UncommonEnums.il) [3](https://github.com/dotnet/runtime/blob/main/src/tests/JIT/opt/IsKnownConstant/TypeGetTypeCodeEnums.il).